haskell/cabal.bzl

```diff
@@ -0,0 +1,247 @@
+        executable = cabal_wrapper,
+        arguments = [args],
+        inputs = depset([setup], transitive = [
+            depset([hs.tools.ghc, hs.tools.ghc_pkg, hs.tools.runghc]),
+            depset(ctx.files.srcs),
+            package_databases,
+            extra_headers,
+            depset(library_deps),
+            depset(ld_library_deps),
+            set.to_depset(dep_info.interface_dirs),
+            depset(dep_info.static_libraries),
+            set.to_depset(dep_info.dynamic_libraries),
+        ]),
+        outputs = [
+            package_database,
+            interfaces_dir,
+            static_library,
+            dynamic_library,
+        ] + ([static_library_prof] if with_profiling else []),
+        env = dicts.add(ghc_env, hs.env),
+        mnemonic = "HaskellCabalLibrary",
+        progress_message = "HaskellCabalLibrary {}".format(hs.label),
+        use_default_shell_env = True,
+    )
+
+    default_info = DefaultInfo(files = depset([static_library, dynamic_library]))
+    hs_info = HaskellInfo(
+        package_ids = set.empty(),
+        package_databases = set.insert(dep_info.package_databases, package_database),
+        version_macros = set.empty(),
+        source_files = set.empty(),
+        extra_source_files = depset(),
+        import_dirs = set.empty(),
+        static_libraries = [static_library] + dep_info.static_libraries,
+        static_libraries_prof = (
+            [static_library_prof] if with_profiling else []
+        ) + dep_info.static_libraries_prof,
+        dynamic_libraries = set.insert(dep_info.dynamic_libraries, dynamic_library),
+        interface_dirs = set.insert(dep_info.interface_dirs, interfaces_dir),
+        compile_flags = [],
+        prebuilt_dependencies = set.empty(),
+        direct_prebuilt_deps = set.empty(),
+        cc_dependencies = dep_info.cc_dependencies,
+        transitive_cc_dependencies = dep_info.transitive_cc_dependencies,
+    )
+    lib_info = HaskellLibraryInfo(package_id = name, version = None)
+    cc_toolchain = find_cpp_toolchain(ctx)
+    feature_configuration = cc_common.configure_features(
+        cc_toolchain = cc_toolchain,
+        requested_features = ctx.features,
+        unsupported_features = ctx.disabled_features,
+    )
+    library_to_link = cc_common.create_library_to_link(
+        actions = ctx.actions,
+        feature_configuration = feature_configuration,
+        dynamic_library = dynamic_library,
+        static_library = static_library,
+        cc_toolchain = cc_toolchain,
+    )
+    compilation_context = cc_common.create_compilation_context()
+    linking_context = cc_common.create_linking_context(
+        libraries_to_link = [library_to_link],
+    )
+    cc_info = cc_common.merge_cc_infos(
+        cc_infos = [
+            CcInfo(
+                compilation_context = compilation_context,
+                linking_context = linking_context,
+            ),
+            cc_info,
+        ],
+    )
```

Comment:
The whole block constructing the `CCInfo` looks extremely boilerplatey, can we factor out the boring parts?